### PR TITLE
Reduce database queries for code.org/learn and /learn/beyond.

### PR DESCRIPTION
**Pull request by breville**
Also set up the commented-out database logging-output line in db.rb so that it won't cause an exception when bin/dashboard-server is run (due to a null $log) and add a comment that it only seems to work when running Pegasus (using bin/pegasus-server).

The single test can be run from pegasus/test with:

```
bundle exec ruby -Itest test_tutorials.rb 
```

And all the tests can be run from there with:

```
bundle exec rake test
```

### CODE REVIEW

**Comment by breville** [comment]
fyi @philbogle @tanyaparker @caleybrock

File: pegasus/src/database.rb [hunk 1]
```diff
@@ -5,10 +5,11 @@ class Tutorials
 
   def initialize(table)
     @table = table
+    @contents = DB[@table].all
   end
 
   def launch_url_for(code,domain)
-    return DB[:beyond_tutorials].where(code: code).first[:url] if @table == :beyond_tutorials
+    return @contents.find {|f| f[:code] == code}[:url] if @table == :beyond_tutorials
```
Comment (laurelfan) [inline pegasus/src/database.rb]: why `f`?

File: pegasus/src/database.rb [hunk 2]
```diff
@@ -5,10 +5,11 @@ class Tutorials
 
   def initialize(table)
     @table = table
+    @contents = DB[@table].all
```
Comment (philbogle) [inline pegasus/src/database.rb:8]: How big will DB[@table].all be for the largest table?

Comment (breville) [inline pegasus/src/database.rb:8]: Good question.  Currently 61 rows, in the form of this csv: https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/cache/Data_HocTutorials.gsheet.csv

File: pegasus/src/database.rb [hunk 3]
```diff
@@ -5,10 +5,11 @@ class Tutorials
...
   end
 
   def launch_url_for(code,domain)
-    return DB[:beyond_tutorials].where(code: code).first[:url] if @table == :beyond_tutorials
+    return @contents.find {|row| row[:code] == code}[:url] if @table == :beyond_tutorials
```
Comment (laurelfan) [inline pegasus/src/database.rb:12]: what else can table be?

Comment (breville) [inline pegasus/src/database.rb:12]: :tutorials

Comment (laurelfan) [inline pegasus/src/database.rb:12]: Is this one still used?
https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/cache/Data_UkHocTutorials.gsheet.csv

**Comment by laurelfan** [comment]
If you're not trying too hard to limit the scope of this change... if you refactored Tutorials to be more of a singleton, I think you could make it so that these tables are only loaded once per process instead of once per page load (similar to how the scripts/levels are loaded). The code in initialize would look more like:

```
@table = table
@@contents_cache ||= Hash.new
@contents = (@@contents_cache[@table] ||= DB[:table])
```

It sounds like the data you have (2 or 3 possible tables, each with <100 rows, doesn't change between deploys) would make this sort of thing make sense. On the other hand you could also argue that the higher level caching in cloudfront makes this unnecessary (because this would only decrease queries by about an order of magnitude since the cloudfront cache expires in 1 hour and we have to restart the pegasus processes every day)

Also... tests ;)? </laurelbot>

**Comment by breville** [comment]
Good idea, but we still want this to be refreshable via page load when on staging, for people editing the gsheets, so will avoid caching for longer than lifetime of a page view.

**Comment by laurelfan** [comment]
LGTM

**Pull request merged**